### PR TITLE
Always check storage access before placing test cookie

### LINKED ISSUE #22839: Linux Firefox + Keycloak 22.0.1 issue(continuation of issue 21307)
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the [latest nightly release](https://github.com/keycloak/keycloak/releases/tag/nightly)

### Area

authentication

### Describe the bug

This is a follow-on to the issue described in #21307

I have a web application that has a react UI that does keycloak authentication against keycloak 21.1.2.   It was discovered that the same issue occurs on Firefox(Rocky9) but it seems primarily when the app is deployed on docker and the authentication is done against our staging instance of keycloak.  However, this morning my boss had the same issue using Safari while talking to the staging version of our application.  That application is deployed using k8s.  

It was suggested in #21307 that 22.0.x fixes the issue.  I deployed 22.0.1 to see if this in fact fixes it.
While it now gets to the app's landing page, after about 5 seconds it get booted back to before the login screen.  My app presents a page with a "Login" button it it.  click this triggers the whole auth process and afterwards sends it to the app's landing page.

image

On Chromium(linux) this doesn't happen, and on Edge(windows) this doesn't happen either.

I have a java (wildfly + vaadin ) application that uses wildfly's oidc client and works fine and has the whole time.  The problem seems to be centered around the react app. 

### Version

22.0.1

### Expected behavior

The app being authenticated using keycloak should log in as normal.

### Actual behavior

The app landing page is pulled up but after about 5 seconds or so it's kicked out. 

### How to Reproduce?

This seems to be particular this one react application, and only with this Firefox on linux and Safari

I thought maybe it was because the app still had keycloak-js 19.0.1 dependency it's package.json.   I cleared out node_modules and changed this dep to be 22.0.1 to agree with latest keycloak release.  no change. 

### Anything else?

everything worked fine in this configuration before update to 21.x + of keycloak, 20.0.x worked great.

### PR DESCRIPTION
**Pull request by jonkoops**
Changes the 3rd-party cookie support detection by first querying support for storage through the [Storage Access API](https://developer.mozilla.org/en-US/docs/Web/API/Storage_Access_API), then the following will happen:

- If storage access is granted, the test cookie is placed to ensure support actually there (as this API can [return false positives](https://developer.mozilla.org/en-US/docs/Web/API/Document/hasStorageAccess#return_value)).
- If storage access is denied, we can safely assume that we don't have access, and we fall back to the redirect flow.

This also changes the Session Status iframe, so that if storage access is denied, an `error` status is signaled. This prevents accidental changes to the session status from being detected, when if fact we simply cannot access the cookie.

Closes #22839,

### CODE REVIEW

**Comment by jonkoops** [comment]
### Test results

Browser | Scenario | Result
-- | -- | --
Chrome | Default behaviour | Places test cookie, correctly detects suppport for 3rd-party cookies.
Chrome | Storage Access API enabled | Uses Storage Access API, correctly detects support for 3rd party cookies.
Chrome | Third-pary cookie phasout enabled | Places test cookie, correctly detects NO suppport for 3rd-party cookies.
Chrome | Third-pary cookie phasout enabled + Storage Access API enabled | Uses Storage Access API, correctly detects NO support for 3rd party cookies.
Firefox | Default behaviour | Uses Storage Access API, correctly detects NO support for 3rd party cookies.
Firefox | Heuristics disabled | Uses Storage Access API, correctly detects NO support for 3rd party cookies.
Safari | Default behaviour | Uses Storage Access API, correctly detects NO support for 3rd party cookies.

#### Browser versions
- Chrome — 116.0.5845.187
- Firefox — 117.0.1
- Safari 16.6

**Review by pedroigor — APPROVED** [review]
@jonkoops LGTM. Hope we finally managed to solve this :)

AFAIK, those HTML resources are strictly related to how SPAs integrate with the server. If we have the quickstarts passing, I'm fine with it.

It would be nice to also have a second review from @stianst.

**Comment by jonkoops** [comment]
> AFAIK, those HTML resources are strictly related to how SPAs integrate with the server. If we have the quickstarts passing, I'm fine with it.


We actually do not have any quick-starts or tests that ensure that Keycloak works in a 3rd-party context (to my knowledge). And we'd have to test every browser listed above in the aforementioned configurations.

So even though we are testing the happy flow of being on the same TLD+1, that doesn't really tell us anything about cross-domain stuff. Still, it's good to know it doesn't regress.

**Review by edewit — APPROVED** [review]

**Review by mposolda — APPROVED** [review]

**Comment by jonkoops** [comment]
I am going to merge this in as I think this PR has gathered enough approvals. I'll have a backport for it up soon.

**Pull request merged**
